src/main/java/net/helix/pendulum/controllers/RoundViewModel.java

```diff
@@ -372,6 +352,55 @@ public static int getRoundIndex(TransactionViewModel milestoneTransaction) {
         return tips;
     }
 
+    /**
+     *
+     *  This method can be used to check whether a transaction has been finalized.
+     *  As getConfirmedTips is used, only transactions with a 2/3rd majority are considered.
+     *
+     * @param tangle tangle
+     * @param security security level
+     * @param transaction transaction
+     * @return boolean, whether the transaction in question exists in the confirmedTransactions set.
+     * @throws Exception Exception
+     */
+    public boolean isTransactionConfirmed(Tangle tangle, int security, Hash transaction) throws Exception {
+        Set<Hash> confirmedTransactions = getReferencedTransactions(tangle, getConfirmedTips(tangle, security));
+        return confirmedTransactions.contains(transaction);
+    }
+
+    /**
+     *
+     * This method can be used to find parents of a tip set
+     * In our case we mainly use it, specifying a tip-set that is completely referenced by a milestone,
+     * to count transaction.confirmations.
+     *
+     * @param tangle tangle
+     * @param tips tips
+     * @return Set of traversed transaction hashes that are parents of the provided tip set
+     * @throws Exception Exception
+     */
+    public Set<Hash> getReferencedTransactions(Tangle tangle, Set<Hash> tips) throws Exception {
+
+        Set<Hash> transactions = new HashSet<>();
+        final Queue<Hash> nonAnalyzedTransactions = new LinkedList<>(tips);
+        Hash hashPointer;
+        while ((hashPointer = nonAnalyzedTransactions.poll()) != null) {
+            final TransactionViewModel transaction = fromHash(tangle, hashPointer);
+            // has round index been set yet?
+            if (transaction.getRoundIndex() == 0) {
```

Comment:
shouldn't if be inverted? Add trunk and branch in nonAnalyzedTransactions if round is != 0?

Reply:
> shouldn't if be inverted? Add trunk and branch in nonAnalyzedTransactions if round is != 0?

The rationale behind this is, that `transaction.roundIndex` is initially set to 0 (and we want to check new transactions). 
It would be set to the `roundIndex` of the_first_ milestone referencing it here https://github.com/HelixNetwork/pendulum/blob/confirmation-state/src/main/java/net/helix/pendulum/service/milestone/impl/MilestoneTrackerImpl.java#L309.  

From my current impression, that is not the expected behavior, the check would additionally need `&& tx.roundIndex == currentIncomingMilestone.roundIndex`.  This way we would also ensure that confirmations are only incremented on milestones of the _same_ round.

Reply:
> > shouldn't if be inverted? Add trunk and branch in nonAnalyzedTransactions if round is != 0?
> 
> The rationale behind this is, that `transaction.roundIndex` is initially set to 0 (and we want to check new transactions).
> It would be set to the `roundIndex` of the_first_ milestone referencing it here https://github.com/HelixNetwork/pendulum/blob/confirmation-state/src/main/java/net/helix/pendulum/service/milestone/impl/MilestoneTrackerImpl.java#L309.
> 
> From my current impression, that is not the expected behavior, the check would additionally need `&& tx.roundIndex == currentIncomingMilestone.roundIndex`. This way we would also ensure that confirmations are only incremented on milestones of the _same_ round.

So, we either set `roundIndex`, increment `confirmations` when round is on initial value 0 OR we just increment `confirmations`, when `roundIndex` of current analyzed milestone corresponds to `roundIndex` of the tx in question.

Reply:
Was resolved with: https://github.com/HelixNetwork/pendulum/pull/198/commits/5d2f4e21472acbc445ef81e4c197a8386960a0c5#diff-94660a248032b99ad1388051966f29b9R391


- _But_, when I uncomment:
https://github.com/HelixNetwork/pendulum/blob/confirmation-state/src/main/java/net/helix/pendulum/service/milestone/impl/MilestoneTrackerImpl.java#L313

- uncomment: https://github.com/HelixNetwork/pendulum/blob/confirmation-state/src/main/java/net/helix/pendulum/service/API.java#L1778

- and, comment: https://github.com/HelixNetwork/pendulum/blob/confirmation-state/src/main/java/net/helix/pendulum/service/API.java#L1779